persist/fs/commitlog/reader.go

```diff
@@ -0,0 +1,279 @@
+	errCommitLogReaderMissingLogMetadata        = errors.New("commit log reader encountered message missing metadata")
+)
+
+type commitLogReader interface {
+	io.Closer
+
+	// Open opens the commit log for reading
+	Open(filePath string) (time.Time, time.Duration, int, error)
+
+	// Read returns the next key and data pair or error, will return io.EOF at end of volume
+	Read() (m3db.CommitLogSeries, m3db.Datapoint, xtime.Unit, []byte, error)
+}
+
+type reader struct {
+	opts           m3db.DatabaseOptions
+	chunkReader    chunkReader
+	chunkReaderRef *chunkReader
+	sizeBuffer     []byte
+	dataBuffer     []byte
+	info           schema.CommitLogInfo
+	log            schema.CommitLog
+	metadata       schema.CommitLogMetadata
+	metadataLookup map[uint64]m3db.CommitLogSeries
+}
+
+func newCommitLogReader(opts m3db.DatabaseOptions) commitLogReader {
+	r := &reader{
+		opts: opts,
+		chunkReader: chunkReader{
+			buffer: bufio.NewReaderSize(nil, bufferWriteSize),
+		},
+		sizeBuffer:     make([]byte, binary.MaxVarintLen64),
+		metadataLookup: make(map[uint64]m3db.CommitLogSeries),
+	}
+	r.chunkReaderRef = &r.chunkReader
+	return r
+}
+
+func (r *reader) Open(filePath string) (time.Time, time.Duration, int, error) {
+	if r.chunkReader.fd != nil {
+		return timeZero, 0, 0, errCommitLogReaderAlreadyOpen
+	}
+
+	var err error
+	r.chunkReader.fd, err = os.Open(filePath)
+	if err != nil {
+		return timeZero, 0, 0, err
+	}
+
+	r.chunkReader.reset(r.chunkReader.fd)
+	r.info = schema.CommitLogInfo{}
+	if err := r.read(&r.info); err != nil {
+		r.Close()
+		return timeZero, 0, 0, err
+	}
+
+	start := time.Unix(0, r.info.Start)
+	duration := time.Duration(r.info.Duration)
+	index := int(r.info.Index)
+	return start, duration, index, nil
+}
+
+func (r *reader) Read() (
+	series m3db.CommitLogSeries,
+	datapoint m3db.Datapoint,
+	unit xtime.Unit,
+	annotation []byte,
+	resultErr error,
+) {
+	if err := r.read(&r.log); err != nil {
+		resultErr = err
+		return
+	}
+
+	if len(r.log.Metadata) != 0 {
+		if err := proto.Unmarshal(r.log.Metadata, &r.metadata); err != nil {
+			resultErr = err
+			return
+		}
+		r.metadataLookup[r.log.Idx] = m3db.CommitLogSeries{
+			UniqueIndex: r.log.Idx,
+			ID:          r.metadata.Id,
+			Shard:       r.metadata.Shard,
+		}
+	}
+
+	metadata, ok := r.metadataLookup[r.log.Idx]
+	if !ok {
+		resultErr = errCommitLogReaderMissingLogMetadata
+		return
+	}
+
+	series = metadata
+	datapoint = m3db.Datapoint{
+		Timestamp: time.Unix(0, r.log.Timestamp),
+		Value:     r.log.Value,
+	}
+	unit = xtime.Unit(byte(r.log.Unit))
+	annotation = r.log.Annotation
+	return
+}
+
+func (r *reader) read(message proto.Message) error {
+	// Read size of message
+	size, err := binary.ReadUvarint(r.chunkReaderRef)
+	if err != nil {
+		return err
+	}
+
+	// Extend buffer as necessary
+	if len(r.dataBuffer) < int(size) {
+		diff := int(size) - len(r.dataBuffer)
+		for i := 0; i < diff; i++ {
+			r.dataBuffer = append(r.dataBuffer, 0)
```

Comment:
`append(...)` will extend the capacity too as necessary, since diff is calculated we'll always have the exact size available when we take a slice of `r.dataBuffer` to read into on line 161.